Feature/st lib

willhuang1997 opened the pull request:
This PR is to separate the designated streamlit components to a library. The apis that can should be consumed from this library will be `VerticalBlock`, `ElementNodeRenderer` and themes from Streamlit if need be. For full api, visit `frontend/lib/index.ts`.

In order to do this, import paths, github actions, local development flows will change and packaging of the library must happen. 

- ***import paths***
  - import paths will be changed from `src/app` to `app/src`. The same will happen to `lib`.
- ***Github Actions***
  - Nothing super significant will change here. Things will get a little slower because we have to run `make protobuf` and `make frontend-lib` and some directories will be changed.
- ***Local Development flows***
  - yarn
    - The changes here now use [yarn workspaces](https://classic.yarnpkg.com/lang/en/docs/workspaces/). `app` still uses CRA but lib has its own bundling and compilation. 
      - yarn buildLib
        - The way that lib bundles and compiles is through `babel` and `tsc`. `Babel` is so our tyepscript code can be used in the browser by transforming our typescript code into browser javascript. `Tsc` is to emit type declarations so that another person can use our library as typescript. We also do a `cp` command that copies our auto-generated proto files to our library.
      - yarn build
        - Uses the build command from CRA so essentially the same except we have to run `yarn buildLib` first.
    - _**yarn start**_
      - _**This will somewhat similar. You can do `yarn start`; however, you will also have to run `python scripts/watch_frontend_lib.py` or `make watch-frontend-lib` from the streamlit root directory to hot reload the lib code. This occurs because @streamlit/lib is its own library. We shall iterate on this workflow later on.**_
      
`lib` and `app` will also have separate jest configs (though they are pretty similar) and tsconfig.json (typescript checks).

willhuang1997 committed:
Minimal exports for app to consume + VerticalBlock and ElementNodeRenderer (#6623)

This PR is the first milestone to determine the minimal number of exports needed for app to consume + VerticalBlock + ElementNodeRenderer.

Most of the changes are import path changes. Some of the more notable changes to look through are

How the @streamlit/lib is built. It's built through the monorepo config for babel (https://babeljs.io/docs/config-files/#monorepos).
a. The babel process works by using a custom preset that is mainly babel-preset-react-app but slightly modified in order to make sure glide data grid code is es6 compatible and hard coding isTypeScriptEnabled to True.
b. The other keys are adding in the emotion plugin for styled-components, ignoring proto, vendor, test files, and telling babel that frontend is the root directory for babel config.
The actual index.ts that is added to frontend/lib
Some things to note that will be punt down but worked on later:

We also need to look into why protobuf can't be compiled through tsc.

Cleaning frontend/app/package.json
Cleaning frontend/app/yarn.lock
Cleaning frontend/lib/scripts/create.js
Determine what's going wrong with globalStyles typing
Determine what's going wrong with styled-components typing for EmotionTheme

willhuang1997 committed:
Change exports for protobuf to be included in index.ts instead of copied through cp command. (#6635)

Tim noticed that I was copying the protobuf.js and protobuf.d.ts instead of compiling it through babel. That causes problems because then you have to import by doing import { ... } from "@streamlit/lib/dist/proto" which doesn't make any sense.
The reason why I did that was because I was getting an error on first try:

`src/proto.js:4:1 - error TS9005: Declaration emit for this file requires using private name 'Radii'. An explicit type annotation may unblock declaration emit.

4 import * as $protobuf from "protobufjs/minimal";
  ~~~~~~

src/proto.js:4:1 - error TS9005: Declaration emit for this file requires using private name 'Radio'. An explicit type annotation may unblock declaration emit.

4 import * as $protobuf from "protobufjs/minimal";`
I just removed the proto files and did a make protobuf and then tried exporting and that worked, as it should.

willhuang1997 committed:
Simple fix for globalStyles (#6651)

willhuang1997 committed:
Fix typing issues for styled-components (#6652)

There are going to be 2 emotion.d.ts. Our @streamlit/lib has our own theming and then our app will simply just be consuming from @streamlit/lib which makes sense.

willhuang1997 committed:
remove extra app directory in frontend dir (#6653)

The current folder structure is frontend/app/src/app. We should make it so that it's frontend/app/src and remove the extra app directory.
This PR also fixes the audit_license script.

willhuang1997 committed:
Merge branch 'develop' into feature/st-lib

willhuang1997 committed:
Merge branch 'develop' into feature/st-lib

willhuang1997 committed:
Remove unnecessary files

willhuang1997 committed:
Fix naming of directories and files

willhuang1997 committed:
Fix BaseColorPicker directory

willhuang1997 committed:
Add Fixes that should exist in feature/st-lib

willhuang1997 committed:
Fix Eslint and Type-checking for st-lib (#6675)

This PR is to fix eslint for app and lib. This also adds dependencies that are necessary to lib's package.json. This is done through eslint noticing the necessary dependencies. This PR also does some linting and also does a merge from develop as we need the BaseButton refactoring to stop duplication of exports from the root index.ts.

In addition, this PR fixes the type checking github actions by adding new commands.

willhuang1997 committed:
Fix notices (#6676)

willhuang1997 committed:
Clean up audit_frontend_licenses

willhuang1997 committed:
Merge branch 'feature/st-lib' of https://github.com/streamlit/streamlit into feature/st-lib

willhuang1997 committed:
Fix/yarn test (#6681)

This is to fix yarn test. This simply adds new scripts through the yarn workspace command to run app and lib tests.
Currently, the lib testing works by mainly using the same scripts that CRA uses for transpiling typescript to javascript.

There is one other thing to note: It looks like if you were to upgrade to 17.0.1 supposedly for react-dom, it would fix jest tests for app to not hang (facebook/react#20756 (comment)). However, jest was still hanging so I just did a delete MessageChannel within src/setupTests.js and we can remove that when we upgrade to React 18.

willhuang1997 committed:
Fix regex that I accidentally forgot to commit (#6698)

willhuang1997 committed:
Fix/cypress (#6699)

willhuang1997 committed:
Remove unused app dep / dev dep and declarations.d.ts (#6713)

* Remove unused app dep / dev dep and declarations.d.ts

* Remove d3 color resolution (#6714)

* Add dependencies removed from app to lib/package.json

* Add pbjs cli

* Readd notices

willhuang1997 committed:
Add watch lib script (#6730)

Since @streamlit/lib is separated out to be its own separate lib, we need a script to constantly rebuild when the files change.

Will also need to update the contributing wiki after this is merged to develop.

willhuang1997 committed:
Get rid of yarn start warnings and clean Makefile (#6739)

willhuang1997 committed:
Remove streamlit.css. I'm not sure where it came from

willhuang1997 committed:
Merge branch 'develop' into feature/st-lib

willhuang1997 committed:
Fix yarn buildLib

willhuang1997 committed:
Fix eslintrc

willhuang1997 committed:
Fix codeql

willhuang1997 committed:
Add notices and yarn.lock

willhuang1997 commented on frontend/app/src/components/StreamlitDialog/__snapshots__/StreamlitDialog.test.tsx.snap:
Seems like this does belong but not sure why this is saying it's added

willhuang1997 commented on frontend/src/app/components/StreamlitDialog/__snapshots__/StreamlitDialog.test.tsx.snap:
We can see that this snapshot is here deleted

willhuang1997 committed:
Fix glide-data-grid versions

willhuang1997 committed:
Merge branch 'develop' into feature/st-lib

willhuang1997 committed:
Merge branch 'develop' into feature/st-lib

sfc-gh-tconkling commented on .pre-commit-config.yaml:
```diff
@@ -66,10 +66,20 @@ repos:
+        name: Typecheck Lib
+        always_run: true
+        # no emit bc tsconfig.json emitDeclaration key is used
+        entry: yarn --cwd frontend/lib tsc
```
Is this noEmit comment meant for typecheck-app? (The `--noEmit` flag is not passed to this `tsc` invocation)

sfc-gh-tconkling commented on frontend/app/package.json:
```diff
@@ -0,0 +1,157 @@
+    ]
+  },
+  "dependencies": {
+    "@streamlit/lib": "1.21.0",
```
Is this dependency version entry auto-updated by our `scripts/update_version.py` script? We should make sure it is (along with the package version for app and lib). (Not a blocker for the merge, I don't think - let's just make sure to do it in a fast-follow PR.)

willhuang1997 replied:
Yep sounds good. I'll do a fast follow PR probably now since this likely won't get merged before st.chat.

sfc-gh-tconkling commented on frontend/babel.config.json:
```diff
@@ -0,0 +1,30 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```
This is a JSON file, so this comment is technically invalid syntax. Can we remove?

sfc-gh-tconkling commented on frontend/jest/fileTransform.js:
```diff
@@ -0,0 +1,58 @@
+const camelcase = require("camelcase")
+
+// This is a custom Jest transformer turning file imports into filenames.
+// Usually, these files aren't particularly useful in tests so we can safely mock them out
```
Micronit: "Usually, these files aren't particularly useful..." makes it sound like the files might still be useful in some tests, which makes mocking them out entirely sound slightly _unsafe_. Let's instead say something like "File imports aren't used by our tests..."

sfc-gh-tconkling commented on frontend/lib/package.json:
```diff
@@ -0,0 +1,151 @@
+{
+    "name": "@streamlit/lib",
+    "version": "1.21.0",
```
(See previous comment re: auto-updating this version string in our `update_version.py` script)

sfc-gh-tconkling commented on frontend/lib/src/SessionInfo.ts:
```diff
@@ -101,7 +101,17 @@ export class SessionInfo {
     const userInfo = initialize.userInfo as UserInfo
     const environmentInfo = initialize.environmentInfo as EnvironmentInfo
-
+    console.log({
+      appId: hashString(userInfo.installationIdV3 + newSession.mainScriptPath),
+      sessionId: initialize.sessionId,
+      streamlitVersion: environmentInfo.streamlitVersion,
+      pythonVersion: environmentInfo.pythonVersion,
+      installationId: userInfo.installationId,
+      installationIdV3: userInfo.installationIdV3,
+      maxCachedMessageAge: config.maxCachedMessageAge,
+      commandLine: initialize.commandLine,
+      userMapboxToken: config.mapboxToken,
+    })
```
This `console.log` - probably left in accidentally? Can we remove?

sfc-gh-tconkling commented on frontend/lib/src/components/shared/StreamlitMarkdown/StreamlitMarkdown.test.tsx:
```diff
@@ -84,11 +84,11 @@ describe("linkReference", () => {
     const body =
-      "Here's a link containing parentheses [Yikes](http://msdn.microsoft.com/en-us/library/aa752574(VS.85).aspx)"
+      "Here's a link containing parentheses [Yikes](http://msdn.microsoft.com/en-usrary/aa752574(VS.85).aspx)"
     const wrapper = mount(getMarkdownElement(body))
 
     expect(wrapper.find("a").prop("href")).toEqual(
-      "http://msdn.microsoft.com/en-us/library/aa752574(VS.85).aspx"
+      "http://msdn.microsoft.com/en-usrary/aa752574(VS.85).aspx"
```
Nit: these look like casualties of a find-replace run amok :) - not a big deal, but let's change them back

sfc-gh-tconkling commented on frontend/lib/src/hocs/withMapboxToken/MapboxTokenError.tsx:
```diff
@@ -59,7 +59,7 @@ const MapboxTokenError = ({
             <p>
               See{" "}
-              <a href="https://docs.streamlit.io/library/advanced-features/configuration#view-all-configuration-options">
+              <a href="https://docs.streamlit.iorary/advanced-features/configuration#view-all-configuration-options">
```
Find-replace error

sfc-gh-tconkling commented on frontend/lib/src/theme/globalStyles.ts:
```diff
@@ -14,11 +14,13 @@
+import { EmotionTheme } from "src/theme"
 
-export const globalStyles = (theme: EmotionTheme): any => css`
+// This theme should be emotionTheme but needs to be researched more why
+// app doesn't pick up theme properly
+export const globalStyles = (theme: EmotionTheme): SerializedStyles => css`
```
I don't understand what this comment means. "This theme should be emotionTheme" - should it be named "emotionTheme" or typed as `EmotionTheme` or something else? What does it mean that "app doesn't pick up theme properly"?

More generally, how should someone reading this comment do this research?

willhuang1997 replied:
Ah this is actually resolved. I need to remove this.

sfc-gh-tconkling commented on frontend/package.json:
```diff
@@ -1,259 +1,26 @@
-  "name": "streamlit-browser",
-  "version": "1.23.1",
+  "name": "streamlit",
+  "version": "1.23.0",
```
See earlier note re: `update_version.py`. Also, let's make sure that all these version numbers match the current streamlit version (1.23.1)

sfc-gh-tconkling commented on frontend/tsconfig.json:
```diff
@@ -1,22 +1,22 @@
-    "target": "es5"
-  },
-  "include": ["src"]
+    "compilerOptions": {
+      "allowJs": true,
+      "allowSyntheticDefaultImports": true,
+      "baseUrl": ".",
+      "esModuleInterop": true,
+      "experimentalDecorators": true,
+      "forceConsistentCasingInFileNames": true,
+      "isolatedModules": true,
+      "jsx": "react-jsx",
+      "lib": ["dom", "dom.iterable", "esnext"],
+      "module": "esnext",
+      "moduleResolution": "node",
+      "noEmit": true,
+      "noFallthroughCasesInSwitch": true,
+      "resolveJsonModule": true,
+      "skipLibCheck": true,
+      "strict": true,
+      "target": "es5"
+    },
+    "include": ["lib/src", "app/src"]
```
Nit: this file is over-indented (indent should be two spaces, not four)

sfc-gh-tconkling commented on scripts/watch_frontend_lib.py:
```diff
@@ -0,0 +1,79 @@
+from watchdog.observers import Observer
+
+# Directory to monitor
+DIRECTORY = os.path.dirname(__file__).join(["./frontend/lib"])
```
Python's `str.join` function is intended for joining lists of strings into one single string. This is a pretty weird usage of it: we're just concatenating two strings, and it would be clearer to use string interpolation, or os.path.join, or the Path module.

```python
DIRECTORY = os.path.join(os.path.dirname(__file__), "./frontend/lib")
```

sfc-gh-wihuang replied:
So I tried testing this out and it didn't work as the directory that it started checking was: 
`scripts/./frontend/lib`

However, I think what we need is:
`scripts/../frontend/lib` so the new DIRECTORY line would look like this if this is fine:
DIRECTORY = os.path.join(os.path.dirname(__file__), "../frontend/lib")

Can you confirm if this is ok?

tconkling replied:
Yep, that's fine!

Alternately, you could use the Path module and do something like `Path(__file__).parent.parent / "frontend/lib"`

sfc-gh-tconkling reviewed (approved):
Looks good! There are a few nits, but the most important next steps IMO are:

- Ensure that all version strings are up to date (they should all be "1.23.1")
- After this is merged, update the `update_version.py` script to handle these new version string locations

tconkling reviewed (approved):
(Whoops - reapproving with the proper account)

willhuang1997 committed:
Fix nits

willhuang1997 committed:
Fix nits

willhuang1997 committed:
Merge branch 'develop' into feature/st-lib

willhuang1997 committed:
Lint and cleanup

willhuang1997 committed:
Minor cleanup for Block utils as comments got removed

lukasmasuch commented on frontend/app/src/components/StreamlitDialog/__snapshots__/StreamlitDialog.test.tsx.snap:
Can we remove this? See comment above.

lukasmasuch commented on frontend/lib/src/components/elements/Spinner/__snapshots__/Spinner.test.tsx.snap:
Can we remove this? See comment above.

lukasmasuch commented on frontend/lib/src/react-app-env.d.ts:
Why is this file needed?

willhuang1997 replied:
Oops accidentally added. will remove.

sfc-gh-wihuang replied:
Actually I think we need this based on github action failures.

lukasmasuch commented on frontend/lib/scripts/create.js:
```diff
@@ -0,0 +1,130 @@
+
+// this is copied from here and modified: https://github.com/facebook/create-react-app/blob/main/packages/babel-preset-react-app/create.js
+
+// Using custom babel plugin. This is needed for preserving es modules for glide data grid
```
Is this file only required because of glide data grid?

willhuang1997 replied:
Yep this file is needed specifically because of glide data grid. glide-data-grid has an exports field that point to cjs when it’s imported using `require()` syntax and we need to keep es modules.

lukasmasuch replied:
hmm, ok. Do you know what we would need to change in the glide-data-grid to get this working out of the box?

lukasmasuch commented on frontend/lib/src/components/elements/ExceptionElement/__snapshots__/ExceptionElement.test.tsx.snap:
Can we remove this? See comment above.

lukasmasuch reviewed (commented):
Overall LGTM. But I wasn't following this project too closely, so I don't have too much context on this.

Some questions:

1. What exactly will be our new development workflow with this change? E.g. I usually have the following setup: One terminal with `streamlit run ...` and one terminal with `yarn start` in `frontend`. And it will hot reload whenever I do any frontend change.

2. How do I build a distribution with this change? I currently do: 1. `make protobuf` 2. `make frontend` 3. `make distribution` in the Streamlit root folder.

2. This isn't necessarily something specific to this change. But I think the three snapshot files that I have commented on above are kind of broken and we should eventually remove those. I believe there isn't any real value in doing a snapshot on the enzyme mount as done in those cases. But this could also be done in a separate PR.

willhuang1997 commented:
>What exactly will be our new development workflow with this change? E.g. I usually have the following setup: One terminal with streamlit run ... and one terminal with yarn start in frontend. And it will hot reload whenever I do any frontend change.

  one terminal for `streamlit run ...`, one terminal for `yarn start`, one terminal for `make watch-frontend-lib`.
  `make watch-frontend-lib` is not currently perfectly hot-reloading but I will be adding a fast-follow PR to improve that experience. The page browser needs to be reloaded right now manually but it does recompile files automatically on save.

>How do I build a distribution with this change? I currently do: 1. make protobuf 2. make frontend 3. make distribution in the Streamlit root folder.

  I just tried `make package` and the flow you suggested. they both should work.

>This isn't necessarily something specific to this change. But I think the three snapshot files that I have commented on above are kind of broken and we should eventually remove those. I believe there isn't any real value in doing a snapshot on the enzyme mount as done in those cases. But this could also be done in a separate PR.

  I will make a separate PR!

willhuang1997 committed:
Remove accidentally committed file

lukasmasuch commented:
> I will make a separate PR!

I have added a discussion topic on that for our next standup. So no rush to get a PR on that out :) 

> The page browser needs to be reloaded right now manually but it does recompile files automatically on save.

Does this mean that I need to refresh the browser manually to see the change?

willhuang1997 committed:
Revert "Remove accidentally committed file"

This reverts commit 87b7f13b9d872f2a39ecd4168fa79cdb33abe385.

sfc-gh-wihuang commented:
>Does this mean that I need to refresh the browser manually to see the change?

Yes :/. I was planning to add a fast-follow PR to look through tabs with localhost:3000 with [pyautogui](https://pyautogui.readthedocs.io/en/latest/) and refresh it.

lukasmasuch commented:
I'm not too worried about manually refreshing the tab on changes. But I still see some risk that it negatively impacts the overall frontend development flow. E.g. how fast does it compile after a change? Is it about the same as in the current hot reloading? Do we see the full stack traces on frontend exceptions, or only the minified versions? ... 

We definitely should take that into the next standup discussion before merging. And maybe you giving a demo of how local frontend development will work with these changes.

willhuang1997 commented:
>I'm not too worried about manually refreshing the tab on changes. But I still see some risk that it negatively impacts the overall frontend development flow. E.g. how fast does it compile after a change? Is it about the same as in the current hot reloading? Do we see the full stack traces on frontend exceptions, or only the minified versions? ...

We definitely should take that into the next standup discussion before merging. And maybe you giving a demo of how local frontend development will work with these changes.

Sure I can give a demo. 
>E.g. how fast does it compile after a change? Is it about the same as in the current hot reloading?

I would say the compilation time is a good amount slower since you have to build the library which in itself is just slow and expected to be some what slow. I would say the compilation will likely take a 30 seconds - a minute. 

>Do we see the full stack traces on frontend exceptions, or only the minified versions?

Let me test that and get back to you or write up something during standup.

willhuang1997 committed:
Merge branch 'develop' into feature/st-lib

willhuang1997 committed:
Fix yarn lock

willhuang1997 committed:
Fix compilation errors

willhuang1997 committed:
Fix lint

willhuang1997 committed:
Fix notices

willhuang1997 committed:
Remove unnecessary package.json dependencies

willhuang1997 committed:
Fix package.json versions

willhuang1997 committed:
Fix yarn.lock and streamlit dialog snapshots

willhuang1997 committed:
Fix notices

willhuang1997 committed:
Have faster hot reloading and absolute imports for @streamlit/lib (#6961)

<!--
⚠️ BEFORE CONTRIBUTING PLEASE READ OUR CONTRIBUTING GUIDELINES!
https://github.com/streamlit/streamlit/wiki/Contributing
-->

## Describe your changes
- Change imports for `app` to `@streamlit/app/src...` as we removed the `ModuleScopePlugin` which allows us to import out of `app/src`. 
- Change imports for `lib` to absolute imports
- add `tsconfig.dev.json` that is specifically for developement
- change `tsconfig.json` for `lib` and `app`
- change `.eslintrc` to `.eslintrc.js` in order to gain access to `__dirname` to make typescript configs relative to each project

## GitHub Issue Link (if applicable)

## Testing Plan
Manually testing
- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?
yes

https://github.com/streamlit/streamlit/assets/16749069/cc0a5f96-f8cd-487b-b70e-6a4e5823f0e7


---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

willhuang1997 committed:
Fix minor comments from Lukas

kmcgrady commented on frontend/app/craco.config.js:
```diff
@@ -82,23 +91,27 @@ module.exports = {
-        })
+      // Remove ModuleScopePlugin which throws when we try to import something
+      // outside of src/.
+      webpackConfig.resolve.plugins.pop()
```
I would prefer a more specific search to remove by name. See further up in the code.

This will make this more resilient to react-scripts.

willhuang1997 replied:
Will make a separate PR to develop after merging.

willhuang1997 committed:
Merge branch 'develop' into feature/st-lib
